widgetry/src/backend_glow.rs

```diff
@@ -418,180 +476,170 @@ impl PrerenderInnards {
-        Vec::with_capacity(img.pixels().len() * bytes_per_pixel);
-
-    // In order to avoid branching in our shader logic, all shapes are rendered with a texture.
-    // Even "non-textured" styles like Fill::Color, use a "default" no-op (pure white) texture,
-    // which we generate here.
-    formatted_pixel_bytes.append(&mut vec![
-        255;
-        (sprite_length * sprite_height) as usize
-            * bytes_per_pixel
-    ]);
-
-    // OpenGL texture arrays expect each texture's bytes to be contiguous, but it's conventional to
-    // store textures in a grid within a single spritesheet image, where a row and column traverses
-    // multiple sprites.
-    //
-    // For example, if we had 6 textures, A-F, the input spritesheet bytes would be like:
-    // [[AAA, BBB, CCC],
-    //  [AAA, BBB, CCC]
-    //  [AAA, BBB, CCC],
-    //  [DDD, EEE, FFF],
-    //  [DDD, EEE, FFF],
-    //  [DDD, EEE, FFF]]
-    //
-    // Which we need to convert to:
-    // [[AAAAAAAAA],
-    //  [BBBBBBBBB],
-    //  [CCCCCCCCC],
-    //  [DDDDDDDDD],
-    //  [EEEEEEEEE],
-    //  [FFFFFFFFF]]
-    use image::GenericImageView;
-    for y in 0..sprites_per_column {
-        for x in 0..sprites_per_row {
-            let sprite_cell = img.view(
-                x * sprite_length,
-                y * sprite_height,
-                sprite_length,
-                sprite_height,
-            );
-            for p in sprite_cell.pixels() {
-                formatted_pixel_bytes.extend_from_slice(&p.2 .0);
+        info!(
+            "img_size: {}x{}px ({} px), sprite_size: {}x{}px, sprites: {}x{} ({} sprites)",
+            img_width,
+            img_height,
+            img.pixels().len(),
+            sprite_width,
+            sprite_height,
+            sprites_per_row,
+            sprites_per_column,
+            sprite_count
+        );
+
+        let mut texture_bytes: Vec<u8> = Vec::with_capacity(img.pixels().len() * bytes_per_pixel);
+
+        // In order to avoid branching in our shader logic, all shapes are rendered with a texture.
+        // Even "non-textured" styles like Fill::Color, use a "default" no-op (pure white) texture,
+        // which we generate here.
+        texture_bytes.append(&mut vec![
+            255;
+            (sprite_width * sprite_height) as usize
+                * bytes_per_pixel
+        ]);
+
+        use image::GenericImageView;
+        for y in 0..sprites_per_column {
+            for x in 0..sprites_per_row {
+                let sprite_cell = img.view(
+                    x * sprite_width,
+                    y * sprite_height,
+                    sprite_width,
+                    sprite_height,
+                );
+                for p in sprite_cell.pixels() {
+                    texture_bytes.extend_from_slice(&p.2 .0);
+                }
             }
         }
-    }
 
-    // prepare and generate mipmaps
-    unsafe {
-        gl.tex_sub_image_3d(
```

Comment:
I don't think this `gl.tex_sub_image_3d ` method was doing anything useful - I deleted it.

I know so little about OpenGL. Anything in here is basically copy/pasted from examples that I don't fully understand.

Reply:
Ditto -- the same is true for all of the buffer binding in `actually_upload`. :P   Graphics programming is such a hard thing to learn, partly because of lack of printf debugging through the whole stack and because when you get things wrong enough, you just don't see anything at all.

I know nothing at all about the texture calls; I may have figured them out in some previous project, also by slowly modifying examples. So no opinion about `tex_sub_image_3d` -- still works after, so cool. :)